File: sdks/java/src/main/java/do_/mongo/MongoClient.java
package do_.mongo;

import java.io.Closeable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * MongoDB Client - the main entry point for database connections.
 * <p>
 * This client manages connections and provides access to databases.
 * </p>
 *
 * <pre>{@code
 * // Create client from connection string
 * MongoClient client = MongoClient.create("mongodb://localhost:27017");
 *
 * // Get a database
 * MongoDatabase db = client.getDatabase("myapp");
 *
 * // Get a collection
 * MongoCollection<Document> users = db.getCollection("users");
 *
 * // Perform operations
 * users.insertOne(new Document("name", "John"));
 *
 * // Close when done
 * client.close();
 * }</pre>
 */
public class MongoClient implements Closeable {

    private static final Pattern URI_PATTERN = Pattern.compile(
            "^(mongodb(?:\\+srv)?)://" +
                    "(?:([^:@]+)(?::([^@]*))?@)?" +
                    "([^/?]+)" +
                    "(?:/([^?]*))?" +
                    "(?:\\?(.*))?$"
    );

    private final MongoClientSettings settings;
    private final Map<String, MongoDatabase> databases;
    private RpcTransport transport;
    private boolean connected;
    private boolean closed;
    private String defaultDatabase;

    /**
     * Creates a MongoClient with settings.
     *
     * @param settings the client settings
     */
    private MongoClient(MongoClientSettings settings) {
        this.settings = settings;
        this.databases = new HashMap<>();
        this.connected = false;
        this.closed = false;
        this.defaultDatabase = settings.getDefaultDatabase();

        // Parse default database from connection string if not set
        if (this.defaultDatabase == null && settings.getConnectionString() != null) {
            ParsedUri parsed = parseUri(settings.getConnectionString());
            this.defaultDatabase = parsed.database;
        }
    }

    /**
     * Creates a MongoClient from a connection string.
     *
     * @param connectionString the MongoDB connection string
     * @return the client
     */
    public static MongoClient create(String connectionString) {
        MongoClientSettings settings = MongoClientSettings.fromConnectionString(connectionString);
        return new MongoClient(settings);
    }

    /**
     * Creates a MongoClient from settings.
     *
     * @param settings the client settings
     * @return the client
     */
    public static MongoClient create(MongoClientSettings settings) {
        return new MongoClient(settings);
    }

    /**
     * Creates and connects a MongoClient asynchronously.
     *
     * @param connectionString the MongoDB connection string
     * @return a CompletableFuture with the connected client
     */
    public static CompletableFuture<MongoClient> connectAsync(String connectionString) {
        MongoClient client = create(connectionString);
        return client.connectAsync().thenApply(v -> client);
    }

    /**
     * Connects to the database.
     *
     * @return this client
     * @throws MongoConnectionException if connection fails
     */
    public MongoClient connect() {
        if (connected) return this;
        if (closed) throw new MongoException("Client is closed");

        try {
            // Create transport (mock for now, real implementation would use HTTP/WebSocket)
            transport = new MockRpcTransport();

            // Perform handshake
            transport.call("connect", settings.getConnectionString());
            connected = true;
            return this;
        } catch (Exception e) {
            throw new MongoConnectionException("Failed to connect: " + e.getMessage(), e);
        }
    }

    /**
     * Connects to the database asynchronously.
     *
     * @return a CompletableFuture
     */
    public CompletableFuture<Void> connectAsync() {
        if (connected) return CompletableFuture.completedFuture(null);
        if (closed) return CompletableFuture.failedFuture(new MongoException("Client is closed"));

        return CompletableFuture.supplyAsync(() -> {
            connect();
            return null;
        });
    }

    /**
     * Gets a database by name.
     *
     * @param name the database name
     * @return the database
     */
    public MongoDatabase getDatabase(String name) {
        ensureConnected();
        return databases.computeIfAbsent(name, n -> new MongoDatabase(transport, n));
    }

    /**
     * Gets the default database (from connection string).
     *
     * @return the default database, or test if not specified
     */
    public MongoDatabase getDatabase() {
        return getDatabase(defaultDatabase != null ? defaultDatabase : "test");
    }

    /**
     * Lists all database names.
     *
     * @return a list of database names
     */
    @SuppressWarnings("unchecked")
    public List<String> listDatabaseNames() {
        ensureConnected();
        Object result = transport.call("listDatabases");
        List<String> names = new ArrayList<>();
        if (result instanceof Document) {
            Object dbs = ((Document) result).get("databases");
            if (dbs instanceof List) {
                for (Object db : (List<?>) dbs) {
                    if (db instanceof Document) {
                        names.add(((Document) db).getString("name"));
                    } else if (db instanceof Map) {
                        names.add((String) ((Map<?, ?>) db).get("name"));
                    }
                }
            }
        } else if (result instanceof Map) {
            Object dbs = ((Map<?, ?>) result).get("databases");
            if (dbs instanceof List) {
                for (Object db : (List<?>) dbs) {
                    if (db instanceof Map) {
                        names.add((String) ((Map<?, ?>) db).get("name"));
                    }
                }
            }
        }
        return names;
    }

    /**
     * Lists all databases with full information.
     *
     * @return a list of database info documents
     */
    @SuppressWarnings("unchecked")
    public List<Document> listDatabases() {
        ensureConnected();
        Object result = transport.call("listDatabases");
        List<Document> databases = new ArrayList<>();
        if (result instanceof Document) {
            Object dbs = ((Document) result).get("databases");
            if (dbs instanceof List) {
                for (Object db : (List<?>) dbs) {
                    if (db instanceof Document) {
                        databases.add((Document) db);
                    } else if (db instanceof Map) {
                        databases.add(new Document((Map<String, Object>) db));
                    }
                }
            }
        } else if (result instanceof Map) {
            Object dbs = ((Map<?, ?>) result).get("databases");
            if (dbs instanceof List) {
                for (Object db : (List<?>) dbs) {
                    if (db instanceof Map) {
                        databases.add(new Document((Map<String, Object>) db));
                    }
                }
            }
        }
        return databases;
    }

    /**
     * Drops a database.
     *
     * @param databaseName the database name
     */
    public void dropDatabase(String databaseName) {
        ensureConnected();
        transport.call("dropDatabase", databaseName);
        databases.remove(databaseName);
    }

    /**
     * Pings the server.
     *
     * @return true if server is reachable
     */
    public boolean ping() {
        ensureConnected();
        try {
            Object result = transport.call("ping");
            if (result instanceof Document) {
                return ((Document) result).getInteger("ok", 0) == 1;
            } else if (result instanceof Map) {
                Object ok = ((Map<?, ?>) result).get("ok");
                return ok instanceof Number && ((Number) ok).intValue() == 1;
            }
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Gets the server status.
     *
     * @return the server status
     */
    @SuppressWarnings("unchecked")
    public Document getServerStatus() {
        ensureConnected();
        Object result = transport.call("serverStatus");
        if (result instanceof Document) {
            return (Document) result;
        } else if (result instanceof Map) {
            return new Document((Map<String, Object>) result);
        }
        return new Document();
    }

    /**
     * Checks if the client is connected.
     *
     * @return true if connected
     */
    public boolean isConnected() {
        return connected && !closed;
    }

    /**
     * Closes the client.
     */
    @Override
    public void close() {
        if (closed) return;
        closed = true;
        connected = false;
        if (transport != null) {
            transport.close();
            transport = null;
        }
        databases.clear();
    }

    /**
     * Gets the settings.
     *
     * @return the client settings
     */
    public MongoClientSettings getSettings() {
        return settings;
    }

    /**
     * Sets a custom transport (for testing).
     *
     * @param transport the transport to use
     */
    public void setTransport(RpcTransport transport) {
        this.transport = transport;
        this.connected = true;
    }

    /**
     * Gets the transport (for testing).
     *
     * @return the transport
     */
    public RpcTransport getTransport() {
        return transport;
    }

    /**
     * Ensures the client is connected.
     */
    private void ensureConnected() {
        if (closed) {
            throw new MongoException("Client is closed");
        }
        if (!connected) {
            connect();
        }
    }

    /**
     * Parses a MongoDB connection URI.
     */
    private ParsedUri parseUri(String uri) {
        ParsedUri parsed = new ParsedUri();
        Matcher matcher = URI_PATTERN.matcher(uri);
        if (matcher.matches()) {
            parsed.protocol = matcher.group(1);
            parsed.username = matcher.group(2);
            parsed.password = matcher.group(3);
            parsed.host = matcher.group(4);
            parsed.database = matcher.group(5);
            parsed.options = matcher.group(6);
        }
        return parsed;
    }

    /**
     * Parsed URI components.
     */
    private static class ParsedUri {
        String protocol;
        String username;
        String password;
        String host;
        String database;
        String options;
    }
}

File: sdks/kotlin/src/main/kotlin/do_/mongo/MongoClient.kt
package do_.mongo

import java.io.Closeable
import java.util.concurrent.ConcurrentHashMap

/**
 * MongoDB Client - the main entry point for database connections.
 *
 * Example usage:
 * ```kotlin
 * // Create client from connection string
 * val client = MongoClient.create("mongodb://localhost:27017")
 *
 * // Or using DSL
 * val client = mongoClient {
 *     connectionString = "mongodb://localhost:27017"
 *     database = "myapp"
 * }
 *
 * // Get a database
 * val db = client.getDatabase("myapp")
 *
 * // Get a collection
 * val users = db.getCollection<Document>("users")
 *
 * // Perform operations
 * users.insertOne(document { "name" to "John" })
 *
 * // Close when done
 * client.close()
 *
 * // Or use 'use' extension
 * MongoClient.create("mongodb://localhost:27017").use { client ->
 *     // operations
 * }
 * ```
 */
class MongoClient private constructor(
    private val settings: MongoClientSettings
) : Closeable {

    private val databases = ConcurrentHashMap<String, MongoDatabase>()
    private var transport: RpcTransport? = null
    private var connected = false
    private var closed = false
    private var defaultDatabase: String? = settings.defaultDatabase

    init {
        // Parse default database from connection string if not set
        if (defaultDatabase == null && settings.connectionString != null) {
            defaultDatabase = parseDatabase(settings.connectionString)
        }
    }

    /**
     * Creates and connects to the database.
     */
    fun connect(): MongoClient {
        if (connected) return this
        if (closed) throw MongoException("Client is closed")

        try {
            transport = MockRpcTransport()
            transport?.call("connect", settings.connectionString)
            connected = true
            return this
        } catch (e: Exception) {
            throw MongoConnectionException("Failed to connect: ${e.message}", e)
        }
    }

    /**
     * Creates and connects to the database asynchronously.
     */
    suspend fun connectAsync(): MongoClient {
        if (connected) return this
        if (closed) throw MongoException("Client is closed")

        try {
            transport = MockRpcTransport()
            transport?.callAsync("connect", settings.connectionString)
            connected = true
            return this
        } catch (e: Exception) {
            throw MongoConnectionException("Failed to connect: ${e.message}", e)
        }
    }

    /**
     * Gets a database by name.
     */
    fun getDatabase(name: String): MongoDatabase {
        ensureConnected()
        return databases.getOrPut(name) { MongoDatabase(transport!!, name) }
    }

    /**
     * Gets the default database (from connection string).
     */
    fun getDatabase(): MongoDatabase = getDatabase(defaultDatabase ?: "test")

    /**
     * Lists all database names.
     */
    @Suppress("UNCHECKED_CAST")
    fun listDatabaseNames(): List<String> {
        ensureConnected()
        val result = transport?.call("listDatabases")
        return extractDatabaseNames(result)
    }

    /**
     * Lists all database names asynchronously.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun listDatabaseNamesAsync(): List<String> {
        ensureConnected()
        val result = transport?.callAsync("listDatabases")
        return extractDatabaseNames(result)
    }

    /**
     * Lists all databases with full info.
     */
    @Suppress("UNCHECKED_CAST")
    fun listDatabases(): List<Document> {
        ensureConnected()
        val result = transport?.call("listDatabases")
        return extractDatabases(result)
    }

    /**
     * Lists all databases with full info asynchronously.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun listDatabasesAsync(): List<Document> {
        ensureConnected()
        val result = transport?.callAsync("listDatabases")
        return extractDatabases(result)
    }

    /**
     * Drops a database.
     */
    fun dropDatabase(databaseName: String) {
        ensureConnected()
        transport?.call("dropDatabase", databaseName)
        databases.remove(databaseName)
    }

    /**
     * Drops a database asynchronously.
     */
    suspend fun dropDatabaseAsync(databaseName: String) {
        ensureConnected()
        transport?.callAsync("dropDatabase", databaseName)
        databases.remove(databaseName)
    }

    /**
     * Pings the server.
     */
    fun ping(): Boolean {
        ensureConnected()
        return try {
            val result = transport?.call("ping")
            when (result) {
                is Document -> result.getInt("ok") == 1
                is Map<*, *> -> (result["ok"] as? Number)?.toInt() == 1
                else -> true
            }
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Pings the server asynchronously.
     */
    suspend fun pingAsync(): Boolean {
        ensureConnected()
        return try {
            val result = transport?.callAsync("ping")
            when (result) {
                is Document -> result.getInt("ok") == 1
                is Map<*, *> -> (result["ok"] as? Number)?.toInt() == 1
                else -> true
            }
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Gets the server status.
     */
    @Suppress("UNCHECKED_CAST")
    fun getServerStatus(): Document {
        ensureConnected()
        val result = transport?.call("serverStatus")
        return when (result) {
            is Document -> result
            is Map<*, *> -> Document(result as Map<String, Any?>)
            else -> Document()
        }
    }

    /**
     * Checks if the client is connected.
     */
    val isConnected: Boolean get() = connected && !closed

    /**
     * Gets the settings.
     */
    fun getSettings(): MongoClientSettings = settings

    /**
     * Sets a custom transport (for testing).
     */
    fun setTransport(transport: RpcTransport) {
        this.transport = transport
        this.connected = true
    }

    /**
     * Gets the transport (for testing).
     */
    fun getTransport(): RpcTransport? = transport

    override fun close() {
        if (closed) return
        closed = true
        connected = false
        transport?.close()
        transport = null
        databases.clear()
    }

    private fun ensureConnected() {
        if (closed) throw MongoException("Client is closed")
        if (!connected) connect()
    }

    @Suppress("UNCHECKED_CAST")
    private fun extractDatabaseNames(result: Any?): List<String> {
        val dbs = when (result) {
            is Document -> result["databases"]
            is Map<*, *> -> result["databases"]
            else -> null
        }
        return when (dbs) {
            is List<*> -> dbs.mapNotNull { db ->
                when (db) {
                    is Document -> db.getString("name")
                    is Map<*, *> -> db["name"]?.toString()
                    else -> null
                }
            }
            else -> emptyList()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun extractDatabases(result: Any?): List<Document> {
        val dbs = when (result) {
            is Document -> result["databases"]
            is Map<*, *> -> result["databases"]
            else -> null
        }
        return when (dbs) {
            is List<*> -> dbs.mapNotNull { db ->
                when (db) {
                    is Document -> db
                    is Map<*, *> -> Document(db as Map<String, Any?>)
                    else -> null
                }
            }
            else -> emptyList()
        }
    }

    private fun parseDatabase(connectionString: String): String? {
        val regex = Regex("^mongodb(?:\\+srv)?://[^/]+/([^?]+)")
        return regex.find(connectionString)?.groupValues?.getOrNull(1)
    }

    companion object {
        /**
         * Creates a MongoClient from a connection string.
         */
        fun create(connectionString: String): MongoClient {
            val settings = MongoClientSettings(connectionString = connectionString)
            return MongoClient(settings)
        }

        /**
         * Creates a MongoClient from settings.
         */
        fun create(settings: MongoClientSettings): MongoClient {
            return MongoClient(settings)
        }

        /**
         * Creates and connects a MongoClient asynchronously.
         */
        suspend fun connectAsync(connectionString: String): MongoClient {
            return create(connectionString).connectAsync()
        }
    }
}

/**
 * Client settings.
 */
data class MongoClientSettings(
    val connectionString: String? = null,
    val host: String = "localhost",
    val port: Int = 27017,
    val defaultDatabase: String? = null,
    val username: String? = null,
    val password: String? = null,
    val authDatabase: String? = null,
    val connectTimeoutMs: Long = 10000,
    val socketTimeoutMs: Long = 0,
    val maxPoolSize: Int = 100,
    val minPoolSize: Int = 0,
    val retryWrites: Boolean = true,
    val retryReads: Boolean = true
) {
    companion object {
        fun fromConnectionString(connectionString: String): MongoClientSettings {
            return MongoClientSettings(connectionString = connectionString)
        }
    }
}

/**
 * DSL builder for MongoClient.
 */
@DocumentDsl
class MongoClientBuilder {
    var connectionString: String? = null
    var host: String = "localhost"
    var port: Int = 27017
    var database: String? = null
    var username: String? = null
    var password: String? = null
    var authDatabase: String? = null
    var connectTimeoutMs: Long = 10000
    var socketTimeoutMs: Long = 0
    var maxPoolSize: Int = 100
    var minPoolSize: Int = 0
    var retryWrites: Boolean = true
    var retryReads: Boolean = true

    fun build(): MongoClient {
        val settings = MongoClientSettings(
            connectionString = connectionString,
            host = host,
            port = port,
            defaultDatabase = database,
            username = username,
            password = password,
            authDatabase = authDatabase,
            connectTimeoutMs = connectTimeoutMs,
            socketTimeoutMs = socketTimeoutMs,
            maxPoolSize = maxPoolSize,
            minPoolSize = minPoolSize,
            retryWrites = retryWrites,
            retryReads = retryReads
        )
        return MongoClient.create(settings)
    }
}

/**
 * Creates a MongoClient using DSL syntax.
 */
inline fun mongoClient(block: MongoClientBuilder.() -> Unit): MongoClient =
    MongoClientBuilder().apply(block).build()

/**
 * Creates a MongoClient and executes a block, closing the client afterwards.
 */
inline fun <T> withMongoClient(connectionString: String, block: (MongoClient) -> T): T {
    return MongoClient.create(connectionString).use { client ->
        client.connect()
        block(client)
    }
}

/**
 * Creates a MongoClient and executes a suspend block, closing the client afterwards.
 */
suspend inline fun <T> withMongoClientAsync(connectionString: String, block: (MongoClient) -> T): T {
    val client = MongoClient.connectAsync(connectionString)
    return try {
        block(client)
    } finally {
        client.close()
    }
}
